// lib-avatar/eedomus.js
const request = require('request');

let _eeDomusConf;

var getPeriphCaract = function (id) {
  return new Promise((resolve, reject) => {

    let url = 'http://'+_eeDomusConf.eeIP+'/api/get?action=periph.caract';
  	url += '&api_user='+_eeDomusConf.eeUser;
  	url += '&api_secret='+_eeDomusConf.eeSecret;
  	url += '&periph_id='+id;

  	request({ 'uri': url, 'json': true, 'encoding': 'binary' }, function (err, response, json){
      if (err || response.statusCode != 200) {
        console.log('getPeriphCaract erreur', err);
  		  return reject(false);
  		}
  		json = json.body;
  		resolve (json);
  	});
  })
}


var getPeriphValueList = function (id) {
  return new Promise((resolve, reject) => {

    let url = 'http://'+_eeDomusConf.eeIP+'/api/get?action=periph.value_list';
  	url += '&api_user='+_eeDomusConf.eeUser;
  	url += '&api_secret='+_eeDomusConf.eeSecret;
  	url += '&periph_id='+id;

  	request({ 'uri': url, 'json': true, 'encoding': 'binary'}, function (err, response, json){
  		if (err || response.statusCode != 200) {
  		  return reject(false);
  		}
  		resolve (json.body);
  	});
  })
}


var set = function (id, value) {
	return new Promise((resolve, reject) => {
			let url = 'http://'+_eeDomusConf.eeIP+'/api/set?action=periph.value&periph_id='+id+'&value='+value+'&api_user='+_eeDomusConf.eeUser+'&api_secret='+_eeDomusConf.eeSecret;
			request({ 'uri': url, 'json': true }, function (err, response, json){
					if (err || response.statusCode != 200)
						  return reject(err);
					resolve();
			});
	})
}


var macro = function (macro_id) {
  return new Promise((resolve, reject) => {
      let url = 'http://'+_eeDomusConf.eeIP+'/api/set?action=periph.macro&macro='+macro_id+'&api_user='+_eeDomusConf.eeUser+'&api_secret='+_eeDomusConf.eeSecret;
      request({ 'uri': url, 'json': true }, function (err, response, json){
          if (err || response.statusCode != 200)
              return reject(err);
          resolve();
      });
  })
}


var getPeriphInfos = function () {
  return new Promise((resolve, reject) => {

    // URL pour la box domotique EEDomus
    let url = 'http://'+_eeDomusConf.eeIP+'/api/get?action=periph.list&notes=1';
  	url += '&api_user='+_eeDomusConf.eeUser;
  	url += '&api_secret='+_eeDomusConf.eeSecret;

    // Requete HTTP pour récupérer la valeur
  	request({ 'uri': url, 'json': true, 'encoding': 'binary'}, function (err, response, json){
  		if (err || response.statusCode != 200) {
  		  return reject ('La récupération de la liste des périphériques a échouée');
  		}
      resolve(json.body);
  	});
  });

}

var EEDomusManager = {

  'init': function(conf) {
		_eeDomusConf = conf ?
									conf :  {
														eeUser: Config.modules.eeDomus.API.User || '',
														eeSecret : Config.modules.eeDomus.API.Secret || '',
														eeIP : Config.modules.eeDomus.API.IP || ''
													};

		return EEDomusManager;
  },
  'set': set,
  'macro': macro,
  'getPeriphCaract': getPeriphCaract,
	'getPeriphValueList': getPeriphValueList,
  'getPeriphInfos': getPeriphInfos
}

// Exports Config
exports.init = EEDomusManager.init;
